Simplified installation instructions (#316)
holochain/docs-pages: src/install.md

You should see something like:

```
nix-shell (Nix) 2.6.0
```

If you’d like to know more about Nix and why we use it, you can [find more information here](../install-advanced/#more-info-on-nix).

---

## Configure Cachix to use pre-built binaries of Holochain

Now that you have installed Nix, you can install and run a development shell that contains all the prerequisites, including the correct Rust and Nodejs versions and the Holochain tools. This shell won’t interfere with your current system configuration.

To significantly speed up the load times for the next step you can make use of our Cachix instance.
If you don't use Cachix, nix-shells will take around 20-30 min for every version of Holochain, depending on your machine's specifications, because it needs to compile the Holochain binaries.

Run the following commands to set up the cache:

```bash
nix-env -iA cachix -f https://cachix.org/api/v1/install
cachix use holochain-ci
```

## Using the Holochain dev tools

> You can use the Holochain dev tools in two ways, either by creating a nix file which specifies the Holochain version or by using a link that points to a current version. Usually Holochain projects define a specific Holochain version that they're running on and compatible with, just like Rust crates or NPM packages are included in a project with a particular version. If you  want play around with Holochain and the dev tools, you can use the link with an ad-hoc nix-shell.

Reviewer: Oh! On re-reading I get what the 'link' method is -- that's the same as holochain.love, right? I was picturing a symlink. In that case, how about this:

```suggestion
You can use the Holochain dev tools in two ways, either by [creating a nix file](#using-holochain-with-a-pinned-holochain-version) which specifies the Holochain version or by using a URL that points to the most recent release. Usually Holochain projects define a specific Holochain version that they're running on and compatible with, just like Rust crates or NPM packages are included in a project with a particular version. If you  want play around with Holochain and the dev tools, you can use the URL with an ad-hoc nix-shell.
```

> You can use the Holochain dev tools in two ways, either by creating a nix file which specifies the Holochain version or by using a link that points to a current version. Usually Holochain projects define a specific Holochain version that they're running on and compatible with, just like Rust crates or NPM packages are included in a project with a particular version. If you  want play around with Holochain and the dev tools, you can use the link with an ad-hoc nix-shell.

Author: Yeah, definitely, thanks for the improvement, @pdaoust.
And sure, let's add the whitespace. I didn't take the time to figure out setting up a local dev server to view the generated site, so all your typesetter view's touches are much appreciated 😀

You should see something like this:

```
holochain 0.0.124
```

Once you `exit` the shell you'll be back to your usual system shell, with all Holochain-specific bits cleaned up.

You can re-enter the Holonix shell with the same command you used initially:

```bash
nix-shell https://holochain.love
```

> It will always keep you up to date with the newest stable version of Holochain and the dev tools.

Reviewer: ```suggestion
It will always keep you up to date with the newest stable version of Holochain and the dev tools.

```

### Using Holochain with a pinned Holochain version

Holochain is currently in rapid development, which means newer versions introduce new features and breaking changes. This means that it's likely that the version that you get with `nix-shell https://holochain.love` won't always work with existing hApp repositories or even breaks a hApp you were working on.

To solve this, repositories of hApp projects can use Nix to pin the appropriate Holochain version to work well for that hApp. To do this, the project needs to have a `default.nix` file in the root folder of the repository. You must not be inside the `nix-shell` provided by https://holochain.love. Instead, simply navigate to the folder with the `default.nix` file and run:

```bash
nix-shell
```
This command looks for a `default.nix` file in the current folder and will create the environment as specified.

> #### Example default.nix file

Reviewer: ```suggestion
#### Example default.nix file

```

> This is a minimal `default.nix` file to use a specific version of Holochain:

Reviewer: ```suggestion
This is a minimal `default.nix` file to use a specific version of Holochain:

```

```nix
let

>   holonixPath = (import ./nix/sources.nix).holonix; # points to the current state of the Holochain repository

Author: What a complicated one-liner. Okay, so that command generates a `default.nix` of this kind?

>   holonixPath = (import ./nix/sources.nix).holonix; # points to the current state of the Holochain repository

Reviewer 2: no it doesn't generate a `default.nix` file at all. it creates the files in `nix/` that the `default.nix` you give in the example relies on.

>   holonixPath = (import ./nix/sources.nix).holonix; # points to the current state of the Holochain repository

Author: So you need to run this command and create/copy a default.nix file and then updating the version of Holochain to the current version of branch `main` works by issuing `nix-shell --run "niv update"`?

>   holonixPath = (import ./nix/sources.nix).holonix; # points to the current state of the Holochain repository

Reviewer 2: @jost-s yes that's correct. the nix code in your example references the file `./nix/sources.nix` which is generated by `niv`.

>   holonixPath = (import ./nix/sources.nix).holonix; # points to the current state of the Holochain repository

Author: @steveeJ Thanks. I've updated that section and added the `niv` init. Please check that section again.

Two more questions:
1. I've tried to upgrade the Holochain dependency in `default.nix` with these two steps. I. e. I created a folder, added the `default.nix` file, then ran the `niv` init command and then executed the `niv` update command. The commands were successful, but the `default.nix` file wasn't changed. So I must be missing something.
2. Please check my initial comment:

> @steveeJ Please review. I need your help with the `niv` command. Looking through our conversations I saw another form of `default.nix`:
> 
> ```nix
> let
>   holonixPath = builtins.fetchTarball "https://github.com/holochain/holonix/archive/$(git ls-remote https://github.com/holochain/holonix main | grep -oP '[0-9a-z]{40}').tar.gz";
>   holonix = import (holonixPath) {
>     holochainVersionId = "v0_0_120";
>   };
>   nixpkgs = holonix.pkgs;
> in nixpkgs.mkShell {
>   inputsFrom = [ holonix.shell ];
>   packages = [
>     # additional packages go here
>   ];
> }
> ```
> 
> Would that be even easier than running `niv`, because one only needs to increase the version number?

Is that so?

>   holonixPath = (import ./nix/sources.nix).holonix; # points to the current state of the Holochain repository

Reviewer 2: > 1. (...). but the default.nix file wasn't changed. So I must be missing something.

yes, none of these commands touch the `default.nix`. i tried to explain this in a previous comment here but it disappeared. `niv update` generates the `nix/sources.nix` based on the `nix/sources.json` file, and it does nothing else. by referencing the generated files in the `default.nix` we have access to the latest holonix revision. if the `default.nix` file specifies the `holochainVersionId = "v...."` argument to the holonix import, it will pin holochain to that version regardless of the holonix update. if that argument is either omitted or names a moving target such as `main` or `develop`, it will move with holonix. if omitted, the default on the `main` branch of holonix is `main`, and the default on the develop branch of holonix is `develop`. i hope this isn't surprising :laughing:

IMO the only thing we're missing to make this process smooth is a command that prints a list of available holochain packages and which ones are cached. if you like you can watch me implement it while i explain to you what i'm doing. it might help you understand the connections.

> 2. Would that be even easier than running niv, because one only needs to increase the version number?

this example works once, when generating the file. you probably took this from between an `cat <<<EOF (...) EOF`, which will substitute the shell command for its output. this will not help with subsequent updates.

  holonix = import (holonixPath) {
    holochainVersionId = "v0_0_124"; # specifies the Holochain version
  };
  nixpkgs = holonix.pkgs;
in nixpkgs.mkShell {
  inputsFrom = [ holonix.main ];
  packages = with nixpkgs; [
    # any additional packages needed for this project, e. g. Nodejs
  ];
}

> ```

Reviewer: ```suggestion
```

```

Author: I've added the whitespace, @pdaoust.

> #### Upgrading Holochain in default.nix

Reviewer: ```suggestion
#### Upgrading Holochain in default.nix

```

> All you need to do to upgrade the version of Holochain in your project is run

Reviewer: ```suggestion
All you need to do to upgrade the version of Holochain in your project is run

```

Reviewer: (this is a very cool new tool, by the way. Is it useful to explain what it does -- I assume it updates your `default.nix` and `Cargo.toml` files? `Cargo.lock` too? Would it be necessary to also mention that breaking changes are your own responsibility to fix, or is that something devs will assume?)

```bash
nix-shell -p niv --run "niv update ..."

> ```

Reviewer: ```suggestion
```

```

Reviewer: weird, this was not what I suggested. GitHub is doing weird things... what I wanted was a newline before and after the code block.

Author: Whitespace added.

```bash
nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/nixos-21.11.tar.gz -p niv --run "niv init && niv drop nixpkgs && niv drop niv && niv add -b main holochain/holonix"
```

Executing this command creates a new folder named `nix` with 2 files that `niv` needs to retrieve the revision of the latest Holochain version.

Next you need to add a `default.nix` file as a Nix configuration of your development environment, including Holochain as a dependency.
This is a minimal `default.nix` file with a specific version of Holochain:

```nix
let
  holonixPath = (import ./nix/sources.nix).holonix; # points to the current state of the Holochain repository
  holonix = import (holonixPath) {
    holochainVersionId = "v0_0_124"; # specifies the Holochain version
  };
  nixpkgs = holonix.pkgs;
in nixpkgs.mkShell {
  inputsFrom = [ holonix.main ];
  packages = with nixpkgs; [
    niv
    # any additional packages needed for this project, e. g. Nodejs
  ];
}
```

Now you can enter the Nix shell with your development environment by running the command `nix-shell`. Once it's finished you have the Holochain commands at your disposal.

#### Upgrading to the latest version of Holochain

> When the time has come to upgrade your hApp to the latest version of Holochain, you first have to set the version id in `default.nix`. E. g.

Reviewer 2: i would say that the
1. first thing to do is `niv update`
2. then run `hn-versions` to see what versions are available
3. and then set the `holochainVersionid` accordingly

Author: Yes, now that `hn-versions` is available, I've changed it so.
